### PR TITLE
Start testing MetaDataManager

### PR DESCRIPTION
**Pull request by mgaudet**
With the aim of starting to support tracking metadata, this pull request starts testing the MetaData Manager.

### CODE REVIEW

**Comment by mgaudet** [comment]
This build fails in Travis; It turns out that the build dependencies are wrong; `libj9avl` isn't built by the time `testjit` is being linked. 

I'll dig into that tomorrow (though, we can still do some review).

File: fvtest/compilertest/tests/MetaDataTest.cpp [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+   data.startPC = 10;
+   data.endPC   = 100;
+   data.name    = "test";
+
+   ASSERT_TRUE(manager.insertMetaData(&data));
+   ASSERT_TRUE(manager.containsMetaData(&data));
+
+   auto* metadata = manager.findMetaDataForPC(10);
+   ASSERT_TRUE(NULL != metadata);
+   ASSERT_TRUE(NULL !=  metadata->name);
```
Comment (Leonardo2718) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: I don't think this assert is required (https://github.com/google/googletest/blob/master/googletest/docs/Primer.md#string-comparison). Is there a reason for it?

File: fvtest/compilertest/tests/MetaDataTest.cpp [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+
+   TR::MethodMetaDataPOD data;
+   data.startPC = 10;
+   data.endPC   = 100;
+   data.name    = "test";
+
+   ASSERT_TRUE(manager.insertMetaData(&data));
+   ASSERT_TRUE(manager.containsMetaData(&data));
+
+   auto* metadata = manager.findMetaDataForPC(10);
```
Comment (Leonardo2718) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: I think
```c++
auto* metadata = manager.findMetaDataForPC(data.startPC);
```
is clearer.

Also is there a reason for using `auto*` instead of just `auto`?

File: fvtest/compilertest/tests/MetaDataTest.cpp [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+   ASSERT_TRUE(NULL != metadata);
+   ASSERT_TRUE(NULL !=  metadata->name);
+   ASSERT_STREQ("test", metadata->name);
```
Comment (Leonardo2718) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: Could this just be

```c++
ASSERT_STREQ(data.name, metadata->name);
```
?

File: fvtest/compilertest/tests/MetaDataTest.cpp [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+
+TEST(MetaDataTest, ExactInsertion) {
+   TestEnhancedCodeMetaDataManager manager;
+   ASSERT_TRUE(manager.initializeCodeMetaDataManager());
+
+   // Set up code cache hash table.
+   ASSERT_TRUE(NULL != manager.addFalseCacheInfo(0,1000));
+
+   TR::MethodMetaDataPOD data;
```
Comment (Leonardo2718) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: Would it be useful to have two copies of this object, one for input and one for expected output?

File: fvtest/compilertest/tests/MetaDataTest.cpp [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+
+   ASSERT_TRUE(manager.insertMetaData(&data));
+   ASSERT_TRUE(manager.containsMetaData(&data));
+
+   auto* metadata = manager.findMetaDataForPC(10);
+   ASSERT_TRUE(NULL != metadata);
+   ASSERT_TRUE(NULL !=  metadata->name);
+   ASSERT_STREQ("test", metadata->name);
+
+   ASSERT_TRUE(manager.removeMetaData(&data));
```
Comment (Leonardo2718) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: Could some of these asserts be moved to a separate test case:

```c++
ASSERT_TRUE(manager.insertMetaData(&data));
ASSERT_TRUE(manager.containsMetaData(&data));
ASSERT_TRUE(manager.removeMetaData(&data));
```

?

File: fvtest/compilertest/tests/MetaDataTest.cpp [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+   // Set up code cache hash table.
+   ASSERT_TRUE(NULL != manager.addFalseCacheInfo(0,1000));
+
+   TR::MethodMetaDataPOD data;
+   data.startPC = 10;
+   data.endPC   = 100;
+   data.name    = "test";
+
+   ASSERT_TRUE(manager.insertMetaData(&data));
+   ASSERT_TRUE(manager.containsMetaData(&data));
```
Comment (Leonardo2718) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: This entire snippet looks the same as the one for `TEST(MetaDataTest, ExactInsertion)`. Could it be moved into a test fixture (https://github.com/google/googletest/blob/master/googletest/docs/Primer.md#test-fixtures-using-the-same-data-configuration-for-multiple-tests)?

File: fvtest/compilertest/tests/MetaDataTest.cpp [hunk 7]
```diff
@@ -0,0 +1,159 @@
...
+
+   std::vector<int> test_points = {11, 25, 99};
+   for (auto i : test_points) { 
+      auto* metadata = manager.findMetaDataForPC(i);
+      ASSERT_TRUE(NULL != metadata) << "Didn't find pc" << i;
+      ASSERT_TRUE(NULL !=  metadata->name);
+      ASSERT_STREQ("test", metadata->name);
+   }
```
Comment (Leonardo2718) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: It's not really clear to me what this test cases is actually testing. Could you add a description? Also, have you considered breaking these up into different test cases.

One last more advanced question (mostly just curios), is there a reason for picking this weak combination of equivalence partitioning and boundary-value analysis?

File: fvtest/compilertest/tests/MetaDataTest.cpp [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+   int iter = 0; 
+   for (auto i : test_points) { 
+      auto* metadata = manager.findMetaDataForPC(i);
+      ASSERT_TRUE(NULL != metadata) << "Didn't find pc" << i;
+      ASSERT_TRUE(NULL !=  metadata->name);
+      ASSERT_STREQ(names[iter++], metadata->name);
+      ASSERT_TRUE(manager.removeMetaData(metadata)); 
+      ASSERT_TRUE(NULL == manager.findMetaDataForPC(i));
+   }
+
```
Comment (Leonardo2718) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: Lots of "magic" numbers here (where did they come from?) so it's not really clear what the test case is trying to detect. Maybe used named values instead and add some more messages to each assert?

File: fvtest/compilertest/tests/MetaDataTest.cpp (hunk 2 above)
Comment (mgaudet) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: Agree on the first point. To the second, IMO, `auto*` provides a readability hint.

File: fvtest/compilertest/tests/MetaDataTest.cpp (hunk 1 above)
Comment (mgaudet) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: Honestly, didn't notice that STREQ does the right thing with `NULL`. Will remove.

File: fvtest/compilertest/tests/MetaDataTest.cpp (hunk 3 above)
Comment (mgaudet) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: Agreed.

File: fvtest/compilertest/tests/MetaDataTest.cpp (hunk 2 above)
Comment (Leonardo2718) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: Fair enough.

File: fvtest/compilertest/tests/MetaDataTest.cpp (hunk 5 above)
Comment (mgaudet) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: I can break them into their own test case; your thinking is if I assert on those other test cases, I can then do `insert`, `contains`, etc,  and not assert in a test case not explicitly trying to test `insert` or `contains`?

**Comment by mgaudet** [comment]
Alright -- based on Leo's feedback I'm going to mark this as WIP, and take another stab at this :)

**Comment by Leonardo2718** [comment]
So I had a few general notes here but GitHub decided to get rid of them for some reason.

First, typo in (4712563cb655f32974069c2fd4bb6bfd4ebfc269)[https://github.com/eclipse/omr/pull/832/commits/4712563cb655f32974069c2fd4bb6bfd4ebfc269]: "assering" -> "asserting"

Second, I think that exposing protected members for the sake of testability is completely OK. In Java, this problem is usually solved using reflection but I think this is (much) more elegant.

Lastly, I think it would be good to document test cases a little more with comments and assert messages.

File: fvtest/compilertest/tests/MetaDataTest.cpp (hunk 5 above)
Comment (Leonardo2718) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: Correct. Having the asserts there is not incorrect (sorry for the double negative), strictly speaking, but it does make it a bit more difficult to understand what's being tested.

File: fvtest/compilertest/tests/MetaDataTest.cpp [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+/**
+ * The design of the CodeMetaData manager is such that it expects each code
+ * cache to be added to the CodeMetaDataManager. In order to avoid that
+ * coupling in this test, the below class allows allocating 'fake' homes for
+ * tree elements
+ *
+ * As well, this resets the singleton on destruction.
+ */
+class TestEnhancedCodeMetaDataManager : public TR::CodeMetaDataManager {
```
Comment (0xdaryl) [inline fvtest/compilertest/tests/MetaDataTest.cpp]: `TestCompiler` namespace

File: compiler/runtime/OMRCodeMetaDataManager.cpp [hunk 10]
```diff
@@ -294,6 +295,12 @@ CodeMetaDataManager::insertMetaDataRangeInHash(
       return 1;
       }
 
+   // Don't insert zero sized ranges.
+   if (startPC == endPC)
+      {
+      return 1;
+      }
+
```
Comment (0xdaryl) [inline compiler/runtime/OMRCodeMetaDataManager.cpp]: What's wrong with inserting zero-sized ranges?  Will this code not function?  Suppose someone wants to insert metadata for a particular code address in their runtime (not sure why, but they might have a need for it).  If we require a range of at least one to cover a code range then it should be documented somewhere in the API.

File: fvtest/compilertest/runtime/CodeMetaDataPOD.hpp [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+#ifndef TR_METHOD_METADATAPOD_INCL
+#define TR_METHOD_METADATAPOD_INCL
+
+#include "runtime/TestCodeMetaDataPOD.hpp"
+#include "infra/Annotations.hpp"           // for OMR_EXTENSIBLE
+
+namespace TR
+{
+
+struct OMR_EXTENSIBLE MethodMetaDataPOD : public Test::MethodMetaDataPODConnector
```
Comment (0xdaryl) [inline fvtest/compilertest/runtime/CodeMetaDataPOD.hpp]: namespace Test -> TestCompiler

File: fvtest/compilertest/runtime/TestCodeMetaDataPOD.hpp [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+#ifndef TEST_METHOD_METADATAPOD_INCL
+#define TEST_METHOD_METADATAPOD_INCL
+
+/*
+ * The following #define(s) and typedef(s) must appear before any #includes in this file
+ */
+#ifndef TEST_METHOD_METADATAPOD_CONNECTOR
+#define TEST_METHOD_METADATAPOD_CONNECTOR
+namespace Test { struct MethodMetaDataPOD; }
+namespace Test { typedef Test::MethodMetaDataPOD MethodMetaDataPODConnector; }
```
Comment (0xdaryl) [inline fvtest/compilertest/runtime/TestCodeMetaDataPOD.hpp]: Namespace Test -> TestCompiler for both

File: fvtest/compilertest/runtime/TestCodeMetaDataPOD.hpp [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+#include "infra/Annotations.hpp"  // for OMR_EXTENSIBLE
+#include "runtime/OMRCodeMetaDataPOD.hpp" 
+
+/*
+ * This structure describes the shape of the method meta data information.
+ * It must be a C++ POD object and follow all the rules behind POD formation.
+ * Because its fields may be extracted by a runtime system, its exact layout
+ * shape MUST be preserved.
+ */
+namespace Test
```
Comment (0xdaryl) [inline fvtest/compilertest/runtime/TestCodeMetaDataPOD.hpp]: Namespace Test -> TestCompiler

File: compiler/runtime/OMRCodeMetaDataManager.cpp (hunk 10 above)
Comment (mgaudet) [inline compiler/runtime/OMRCodeMetaDataManager.cpp]: Currently, you can insert metadata with a zero sized range, but the query logic will never find it. So you'll insert things into the table that can never be removed. I'll add a comment to this.

**Comment by mgaudet** [comment]
Addressed comments.
